app_integrations/apps/box.py

```diff
@@ -0,0 +1,211 @@
+        # parameters only if the stream_position parameter is not included.
+        if self._next_stream_position:
+            params['stream_position'] = self._next_stream_position
+        else:
+            params['created_after'] = self._last_timestamp
+
+        def _perform_request(allow_retry=True):
+            try:
+                # Get the events using a make_request call with the box api. This is to
+                # support custom parameters such as 'created_after' and 'created_before'
+                box_response = self._client.make_request(
+                    'GET',
+                    self._client.get_url('events'),
+                    params=params
+                )
+            except BoxException:
+                LOGGER.exception('Failed to get events for %s', self.type())
+                return False
+            except ConnectionError:
+                # In testing, the requests connection seemed to get reset for no
+                # obvious reason, and a simple retry once works fine so catch it
+                # and retry once, but after that return False
+                LOGGER.exception('Bad response received from host, will retry once')
+                if allow_retry:
+                    return _perform_request(allow_retry=False)
+
+                return False
+
+            # Return the JSON from the box response
+            return box_response.json()
+
+        return _perform_request()
+
+    def _gather_logs(self):
+        """Gather the Box Admin Events
+
+        The ideal way to do this would be to use the boxsdk.events.Events class and the
+        `get_events` method to retrieve these events. However, this method does allow you
+        to pass keyword arguments (such as params) which are needed for specifying the
+        'created_after' parameter.
+
+        Returns:
+            bool or list: If the execution fails for some reason, return False.
+                Otherwise, return a list of box admin event entries.
+        """
+        if not self._create_client():
+            return False
+
+        response = self._make_request()
+
+        if not response:
+            LOGGER.error('No results received from the Box API request for %s', self.type())
+            return False
+
+        self._more_to_poll = int(response['chunk_size']) >= self._MAX_CHUNK_SIZE
+
+        events = response.get('entries', [])
+        if not events:
+            LOGGER.info('No events in response from the Box API request for %s', self.type())
+            return False
+
+        self._next_stream_position = response['next_stream_position']
+
+        self._last_timestamp = events[-1]['created_at']
+
+        return events
+
+    @classmethod
+    def _required_auth_info(cls):
+        # Use a validation function to ensure the file the user provides is valid
+        def keyfile_validator(keyfile):
+            """A JSON formatted Box service account private key file key"""
```

Comment:
Looks like some args and return values are missing in here

Reply:
I had, yes, but this isn't an error that's caused by too many requests, etc and is an error that is caused (I think) by a bug in the box sdk where it is not properly refreshing the auth token (again, I think). Here is a sample of the traceback I saw once or twice during testing:
```
Traceback (most recent call last):
  File "/Users/my_user/Desktop/testopen.py", line 205, in gather_logs
    params=params
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/util/api_call_decorator.py", line 71, in call
    return method(*args, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/client/client.py", line 407, in make_request
    return self._session.request(method, url, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 510, in request
    response = self._prepare_and_send_request(method, url, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 350, in _prepare_and_send_request
    **kwargs
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 395, in _make_request
    access_token_will_be_used = self._renew_session(None)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 196, in _renew_session
    new_access_token, _ = self._oauth.refresh(access_token_used)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/oauth2.py", line 251, in refresh
    access_token, refresh_token = self._refresh(access_token_to_refresh)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 394, in _refresh
    new_access_token = self.authenticate_instance()
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 384, in authenticate_instance
    return self._auth_with_jwt(self._enterprise_id, 'enterprise')
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 243, in _auth_with_jwt
    return self._construct_and_send_jwt_auth(sub, sub_type)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 219, in _construct_and_send_jwt_auth
    return self.send_token_request(data, access_token=None, expect_refresh_token=False)[0]
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/oauth2.py", line 359, in send_token_request
    token_response = self._execute_token_request(data, access_token, expect_refresh_token)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/oauth2.py", line 328, in _execute_token_request
    access_token=access_token,
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/network/default_network.py", line 24, in request
    request_response=self._session.request(method, url, **kwargs),
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/requests/adapters.py", line 490, in send
    raise ConnectionError(err, request=request)
ConnectionError: ('Connection aborted.', error(54, 'Connection reset by peer'))
```

I will probably implement backoff in the apps at some point, but will want to do it on broader level and not a per-app one off.

Reply:
@jacknagz that is by design because these docstrings are used to provide more context to the user if validation fails. See the `__doc__` referenced here: https://github.com/airbnb/streamalert/blob/cd9f1b5d5521717e7d01b1bb6c7c294c697af10a/stream_alert_cli/helpers.py#L528-L530